### PR TITLE
[dynamo] Add support for data_ptr() comparisons in torch.compile

### LINKED ISSUE #165408: [torch.compile][fullgraph] Dynamo rejects comparing `data_ptr()` results (`Unsupported: Builtin operator.* comparison with constant self`)
### 🐛 Describe the bug

A tiny model that compares tensor data_ptr()/stride()/shape between an input and its detach() runs fine in eager, but fails under torch.compile(fullgraph=True) in the Dynamo front-end with:
```
torch._dynamo.exc.Unsupported: Builtin `operator.*` comparison with constant `self` failed
Explanation: Failed to compare DataPtrVariable() with DataPtrVariable(), because DataPtrVariable() is not a Python constant or its mutation check fails.
```
This blocks full-graph compilation for otherwise pure-inference code that only inspects metadata and returns a boolean-as-long tensor.

### Minimal Repro
```python
import time, torch
from torch import nn

def env_info():
    print(f"[Env] torch.__version__ = {torch.__version__}")
    print(f"[Env] cuda.is_available = {torch.cuda.is_available()}")
    if torch.cuda.is_available():
        print(f"[Env] torch.version.cuda = {torch.version.cuda}")
        print(f"[Env] torch.backends.cudnn.version() = {torch.backends.cudnn.version()}")

class MyModel(nn.Module):
    def forward(self, x):
        original = x
        detached = x.detach()
        same_data = (
            original.data_ptr() == detached.data_ptr() and
            original.stride() == detached.stride() and
            original.shape == detached.shape
        )
        return torch.tensor([same_data], dtype=torch.long, device=x.device)

def GetInput():
    return torch.rand(1, 1, 1, 3, dtype=torch.float32)

def run_once(model, x, label="run"):
    model.eval()
    if torch.cuda.is_available(): torch.cuda.synchronize()
    t0 = time.perf_counter()
    with torch.no_grad():
        y = model(x)
    if torch.cuda.is_available(): torch.cuda.synchronize()
    print(f"[{label}] output: {y.cpu().tolist()}")
    return y

def main():
    env_info()
    device = "cuda" if torch.cuda.is_available() else "cpu"
    model = MyModel().to(device).eval()
    x = GetInput().to(device)

    # Eager OK
    run_once(model, x, "eager")

    # Compile fullgraph -> fails
    compiled = torch.compile(model, fullgraph=True)
    run_once(compiled, x, "compiled")

if __name__ == "__main__":
    main()
```

### Error logs
```
[eager] output: [1]
...
torch._dynamo.exc.Unsupported: Builtin `operator.*` comparison with constant `self` failed
Explanation: Failed to compare DataPtrVariable() with DataPtrVariable(), because DataPtrVariable() is not a Python constant or its mutation check fails.

from user code:
  original.data_ptr() == detached.data_ptr()
  ...
Set TORCHDYNAMO_VERBOSE=1 / TORCH_LOGS="+dynamo" for full internal trace.
```

### Versions

```
PyTorch version:  2.8.0a0+gitba56102 
Is debug build: True
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 12.3.0-1ubuntu1~22.04) 12.3.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.16 (main, Dec 11 2024, 16:24:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.8.0-59-generic-x86_64-with-glibc2.35
Is CUDA available: True
```

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @amjames @Lucaskabela @jataylo @chenyang78

### PR DESCRIPTION
**Pull request by hinriksnaer**
Fixes #165408

## Summary

This PR enables `data_ptr()` comparisons to work with `torch.compile(fullgraph=True)` by implementing comparison operators (`__eq__`, `__ne__`) for `DataPtrVariable`. Previously, comparing data pointers would cause an `Unsupported` error and graph break.

## Problem

When using `torch.compile` with `fullgraph=True`, code that compares tensor data pointers would fail:

```python
import torch

def check_aliasing(x):
    detached = x.detach()
    # Check if detach shares the same data pointer
    return x.data_ptr() == detached.data_ptr()

model = torch.compile(check_aliasing, fullgraph=True)
x = torch.randn(10)
model(x)  # ❌ torch._dynamo.exc.Unsupported: Builtin `operator.*` comparison with constant `self` failed
```

**Error:**
```
torch._dynamo.exc.Unsupported: Builtin `operator.*` comparison with constant `self` failed
Explanation: Failed to compare DataPtrVariable() with DataPtrVariable(), because DataPtrVariable() is not a Python constant or its mutation check fails.
```

## Root Cause

`DataPtrVariable` did not implement `call_method()` for comparison operators, so the base class `VariableTracker.call_method()` would attempt to treat it as a Python constant and fail.

## Solution

Implemented `call_method()` in `DataPtrVariable` to handle `__eq__` and `__ne__` comparisons by evaluating them at compile time using fake tensors.

## Testing

Run the new test suite:
```bash
python test/dynamo/test_data_ptr.py -v
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela @mlazos

### CODE REVIEW

File: torch/_dynamo/variables/tensor.py [hunk 1]
```diff
@@ -1807,7 +1807,44 @@ def __init__(
...
+    def call_method(
+        self,
+        tx,
+        name,
+        args: "list[VariableTracker]",
+        kwargs: "dict[str, VariableTracker]",
+    ) -> "VariableTracker":
+        if name == "__eq__" and len(args) == 1:
+            if not isinstance(args[0], DataPtrVariable):
+                return variables.ConstantVariable.create(NotImplemented)
```
Comment (Lucaskabela) [inline torch/_dynamo/variables/tensor.py]: Doesn't __eq__ default to false for cases where the other arg is not the same data type?

Comment (Lucaskabela) [reply torch/_dynamo/variables/tensor.py]: I think there is also a positive case when the data_ptr has no data and is 0 (null), i.e

```
x = torch.tensor((), dtype=torch.bool)
x.data_ptr() == 0
```

**Review by Lucaskabela — CHANGES_REQUESTED** [review]
Minor questions how we handle explicitly comparing to int, but this change generally looks good with some clarification needed

File: torch/_dynamo/variables/tensor.py (hunk 1 above)
Comment (hinriksnaer) [reply torch/_dynamo/variables/tensor.py]: Thanks @Lucaskabela  for the detailed review! I am still onboarding onto dynamo so this feedback is much appreciated. I updated the code and refined/expanded the tests based on your feedback. Some tests are auto-generated so I encourage you to check if I missed something while verifying their correctness and coverage.

### Returning NotImplemented vs False for incompatible types

This is a good point to raise and something I did not consider. My assumption would be that returning NotImplemented is the correct Python comparison because it gives Python a chance to try the reverse comparison first (b.__eq__(a)). This follows the pattern used elsewhere in dynamo e.g. variables/base.py:503.

I've added test_data_ptr_comparison_to_int_constant_works which verifies this works correctly - when we return NotImplemented for data_ptr() == 99999, Python's protocol handles it via fallback and the comparison completes successfully.

---

### Empty tensors with data_ptr() == 0

This was a good catch  Empty tensors have null pointers (data_ptr() == 0), so two different empty tensors should compare equal.

Fixed by checking storage size upfront:
```python
  self_has_storage = self_example.untyped_storage().size() > 0
  other_has_storage = other_example.untyped_storage().size() > 0

  if not self_has_storage and not other_has_storage:
      return variables.ConstantVariable.create(True)  # Both have data_ptr == 0

  # (Using untyped_storage().size() instead of data_ptr() because we can't call data_ptr() on fake tensors)
```
---
As I mentioned before, I am pretty new to dynamo so don't hesitate to provide more feedback and insights. I hope this addresses all your points.

File: torch/_dynamo/variables/tensor.py [hunk 2]
```diff
@@ -1814,7 +1814,55 @@ def __init__(
...
+    def call_method(
+        self,
+        tx,
+        name,
+        args: "list[VariableTracker]",
+        kwargs: "dict[str, VariableTracker]",
+    ) -> "VariableTracker":
+        if name == "__eq__" and len(args) == 1:
+            other = args[0]
+            if not isinstance(self, type(other)):
```
Comment (Lucaskabela) [inline torch/_dynamo/variables/tensor.py]: Okay, I am understanding this logic because of the `__eq__` polyfill - seems okay to return NotImplemented here

File: test/dynamo/test_data_ptr.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+        """Zero-sized multi-dimensional tensors have data_ptr() == 0"""
+
+        def fn(x):
+            return x.data_ptr() == x.data_ptr()
+
+        zero_sized = torch.empty(5, 0)
+        opt_fn = torch.compile(fn, fullgraph=True)
+        self.assertTrue(opt_fn(zero_sized))
+        self.assertEqual(fn(zero_sized), opt_fn(zero_sized))
+
```
Comment (Lucaskabela) [inline test/dynamo/test_data_ptr.py:363]: Here is another testcase for us to consider (and shows why we should not be comparing as compile time, but should defer for runtime):

```python
def fn(x, y):
    return x.data_ptr() == y

x = torch.randn(4, 4)
y = torch.randn(4, 4)
self.assertEqual(x, x.data_ptr())
self.assertEqual(y, y.data_ptr())
```

Comment (Lucaskabela) [reply test/dynamo/test_data_ptr.py:363]: To fix this test case, I think we will need to update to properly have DataPtrVariable proxy

File: torch/_dynamo/variables/tensor.py [hunk 4]
```diff
@@ -1814,7 +1814,55 @@ def __init__(
...
+                return variables.ConstantVariable.create(NotImplemented)
+
+            self_example = self.from_tensor.proxy.node.meta.get("example_value")
```
Comment (Lucaskabela) [inline torch/_dynamo/variables/tensor.py]: I don't think we should use `example_value` here - see my test case as example

**Review by Lucaskabela — CHANGES_REQUESTED** [review]
Think there is another edge case to consider here we should get workign

File: torch/_dynamo/variables/tensor.py [hunk 5]
```diff
@@ -1809,12 +1824,69 @@ class DataPtrVariable(VariableTracker):
...
+        name,
+        args: "list[VariableTracker]",
+        kwargs: "dict[str, VariableTracker]",
+    ) -> "VariableTracker":
+        if name == "__eq__" and len(args) == 1:
+            other = args[0]
+            if not isinstance(other, DataPtrVariable):
+                # Comparing a data_ptr (int) to a non-data_ptr value
+                # If comparing to a tensor, it's always False
+                if isinstance(other, TensorVariable):
```
Comment (Lucaskabela) [inline torch/_dynamo/variables/tensor.py]: Why do we need to special case the TensorVariable? Can we just fallback regardless?

File: torch/_dynamo/graph_break_registry.json [hunk 6]
```diff
@@ -2820,5 +2820,13 @@
         "It may be possible to write Dynamo tracing rules for this code. Please report an issue to PyTorch if you encounter this graph break often and it is causing performance issues."
       ]
     }
+  ],
+  "GB0280": [
```
Comment (Lucaskabela) [inline torch/_dynamo/graph_break_registry.json:2856]: Is this still needed?

File: torch/_dynamo/variables/tensor.py (hunk 5 above)
Comment (hinriksnaer) [reply torch/_dynamo/variables/tensor.py]: You're totally right, this is redundant. I will update this and make the linter happy

File: test/dynamo/test_data_ptr.py (hunk 3 above)
Comment (Lucaskabela) [reply test/dynamo/test_data_ptr.py:363]: Sorry I flubbed this testcase - see below

File: test/dynamo/test_data_ptr.py [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+            result = x.data_ptr() == y
+            # Convert to scalar for testing purposes
+            return result.all().item()
+
+        x = torch.randn(4, 4)
+        y = torch.randn(4, 4)
+        opt_fn = torch.compile(fn, fullgraph=True)
+        # data_ptr (int) should never equal any element in tensor
+        self.assertFalse(opt_fn(x, y))
+        self.assertEqual(fn(x, y), opt_fn(x, y))
```
Comment (Lucaskabela) [inline test/dynamo/test_data_ptr.py]: Thanks - I think we are getting closer, but there is still something missing here.  Take for instance this:
```python
import torch
def fn(x: torch.Tensor, y: int) -> bool:
    return x.data_ptr() == y 

x = torch.randn(4, 4)
opt_fn = torch.compile(fn, fullgraph=True)
self.assertEqual(fn(x, x.data_ptr()), opt_fn(x, x.data_ptr()))
```

File: torch/_dynamo/variables/tensor.py [hunk 8]
```diff
@@ -1809,12 +1824,69 @@ class DataPtrVariable(VariableTracker):
...
+                    return variables.ConstantVariable.create(False)
```
Comment (Lucaskabela) [inline torch/_dynamo/variables/tensor.py]: I am not sure falling back is correct to do since we want to insert when possible

File: torch/_dynamo/variables/tensor.py [hunk 9]
```diff
@@ -1809,12 +1824,69 @@ class DataPtrVariable(VariableTracker):
...
+                # For other types (e.g., int constants), return NotImplemented
+                # to allow Python's comparison protocol to handle it
+                return variables.ConstantVariable.create(NotImplemented)
+
+            # Runtime approach: Insert the comparison into the graph
+            from .builder import wrap_fx_proxy
+
+            proxy = tx.output.create_proxy(
```
Comment (Lucaskabela) [inline torch/_dynamo/variables/tensor.py]: So let's focus in on the super simple case enumerated above, where `x.data_ptr() == y`.  When we cal `self.as_proxy()` what happens? 

Within `aot_eager` (in ProxyTensor decomposition), this gets realized into a const value because of our prim op; as a result, we end up with an aot_eager graph that looks like this:

```
I1027 15:34:12.471000 1003987 torch/_functorch/_aot_autograd/graph_capture.py:305] [0/1] [__aot_graphs]  ===== Forward graph 1 =====
I1027 15:34:12.471000 1003987 torch/_functorch/_aot_autograd/graph_capture.py:305] [0/1] [__aot_graphs]  /home/lucaskabela/pytorch/torch/fx/_lazy_graph_module.py class <lambda>(torch.nn.Module):
I1027 15:34:12.471000 1003987 torch/_functorch/_aot_autograd/graph_capture.py:305] [0/1] [__aot_graphs]     def forward(self, arg0_1: "f32[4, 4][4, 1]cpu", arg1_1: "Sym(s48)"):
I1027 15:34:12.471000 1003987 torch/_functorch/_aot_autograd/graph_capture.py:305] [0/1] [__aot_graphs]          # File: /home/lucaskabela/pytorch/test.py:5 in eq_data_ptr, code: return x.data_ptr() == y
I1027 15:34:12.471000 1003987 torch/_functorch/_aot_autograd/graph_capture.py:305] [0/1] [__aot_graphs]         eq: "Sym(Eq(s48, 6000000))" = arg1_1 == 6000000;  arg1_1 = None
I1027 15:34:12.471000 1003987 torch/_functorch/_aot_autograd/graph_capture.py:305] [0/1] [__aot_graphs]         return (eq,)
I1027 15:34:12.471000 1003987 torch/_functorch/_aot_autograd/graph_capture.py:305] [0/1] [__aot_graphs]         
```

This will evaluate to false and produce incorrect output; but we are on the right path! 

What we want to do for our `proxy` is call the `data_ptr` function on the tensor; maybe we need to modify `as_proxy` to work like that

**Review by Lucaskabela — CHANGES_REQUESTED** [review]
Thanks for iterating on this and the patience - I think we are close to a working solution.  Suggested a slightly better testcase and some changes

File: torch/_dynamo/variables/tensor.py (hunk 9 above)
Comment (hinriksnaer) [reply torch/_dynamo/variables/tensor.py]: So this is definitely tricky to implement and I might need some time to get more familiar with the internals. In the meantime, any additional suggestions or pointers on implementation details would be much appreciated.

**Comment by hinriksnaer** [comment]
New set of changes and new tests should all pass now.
 
  ```python
  def fn(x, ptr_value):
      return x.data_ptr() == ptr_value  # Now works
```
This was however a bigger challenge than it might have seemed in the beginning since getting runtime values requires a decent amount of workaround given the current setup. This is a working solution but I could use some expertise to determine if this involves too many workarounds. In any case, there are three options the way I see it

1. Merge this change
2. Merge this change + scope in a more elegant "long term" solution
3. Merge a smaller change where we still fail some edge cases + scope in additional work required for full coverage

Let me know what you think and if there is anything worth looking into. It would be good to get this out without extending the current scope too much.

File: torch/_dynamo/data_ptr_op.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+import weakref
+
+import torch
+from torch._prims import _make_prim, RETURN_TYPE
+from torch._subclasses import FakeTensorMode
+from torch._subclasses.functional_tensor import FunctionalTensorMode
+
+
+# Fake pointer base value (high value to avoid collision with user constants)
+_FAKE_PTR_BASE = 1_000_000
```
Comment (Lucaskabela) [inline torch/_dynamo/data_ptr_op.py]: I am a tad worried with this - hardcoding some ptr value opens us up to a "heisenbug" where one day we could incidentally hit this value on accident

File: torch/_dynamo/data_ptr_op.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
+"""Data pointer operations for Dynamo tracing.
+
+This module provides primitives for handling tensor.data_ptr() during tracing. It includes:
```
Comment (Lucaskabela) [inline torch/_dynamo/data_ptr_op.py:3]: I am not 100% convinced this is the best approach here (I am not familiar with how we typically do this tbh), so I want @bdhirsh  to take a look if possible

File: torch/_dynamo/data_ptr_op.py (hunk 10 above)
Comment (hinriksnaer) [reply torch/_dynamo/data_ptr_op.py]: Fixed

**Comment by bdhirsh** [comment]
I'm not sure how far we should try to bend over backward to natively support data_ptr() access in user code during graph capture.

(1) Most usages of data_ptr (that I'm aware of at least) are mainly for plumbing data pointers into custom kernels. The usual play to make this compilable is to hide any data_ptr usage in a custom op (which the user authors, and doesn't need to live in pytorch proper

(2) at the very least, these new custom ops as written right now will burn an `int` into the graph which is not correct (we don't have real data_ptrs at trace time to put into the graph that are representative of what the runtime tensor data pointers are)

cc @zou3519 for thoughts

**Review by zou3519 — COMMENTED** [review]
We're not going to bend over backward to support data_ptr in Dynamo unless there's a good reason for it.
